ouroboros-network/protocols/lib/Ouroboros/Network/Protocol/ObjectDiffusion/Codec.hs

```diff
@@ -0,0 +1,287 @@
+timeLimitsObjectDiffusion = ProtocolTimeLimits stateToLimit
+  where
+    stateToLimit
+      :: forall (st :: ObjectDiffusion objectId object).
+         (ActiveState st)
+      => StateToken st
+      -> Maybe DiffTime
+    stateToLimit SingInit                        = waitForever
+    stateToLimit (SingObjectIds SingBlocking)    = waitForever
+    stateToLimit (SingObjectIds SingNonBlocking) = shortWait
+    stateToLimit SingObjects                     = shortWait
+    stateToLimit SingIdle                        = waitForever
+    stateToLimit a@SingDone                      = notActiveState a
+
+codecObjectDiffusion
+  :: forall (objectId :: Type) (object :: Type) m.
+     (MonadST m)
+  => (objectId -> CBOR.Encoding)         -- ^ encode 'objectId'
+  -> (forall s. CBOR.Decoder s objectId) -- ^ decode 'objectId'
+  -> (object   -> CBOR.Encoding)         -- ^ encode object
+  -> (forall s. CBOR.Decoder s object)   -- ^ decode object
+  -> Codec (ObjectDiffusion objectId object) CBOR.DeserialiseFailure m ByteString
+codecObjectDiffusion encodeObjectId decodeObjectId encodeObject decodeObject =
+  mkCodecCborLazyBS
+    (encodeObjectDiffusion encodeObjectId encodeObject)
+    decode
+    where
+      decode
+        :: forall (st :: ObjectDiffusion objectId object).
+           (ActiveState st)
+        => StateToken st
+        -> forall s. CBOR.Decoder s (SomeMessage st)
+      decode stok = do
+        len <- CBOR.decodeListLen
+        key <- CBOR.decodeWord
+        decodeObjectDiffusion decodeObjectId decodeObject stok len key
+
+encodeObjectDiffusion
+  :: forall (objectId :: Type) (object :: Type)
+            (st  :: ObjectDiffusion objectId object)
+            (st' :: ObjectDiffusion objectId object).
+     (objectId -> CBOR.Encoding) -- ^ encode 'objectId'
+  -> (object   -> CBOR.Encoding) -- ^ encode 'object'
+  -> Message (ObjectDiffusion objectId object) st st'
+  -> CBOR.Encoding
+encodeObjectDiffusion encodeObjectId encodeObject = encode
+  where
+    encode
+      :: forall st0 st1.
+         Message (ObjectDiffusion objectId object) st0 st1
+      -> CBOR.Encoding
+    encode MsgInit =
+         CBOR.encodeListLen 1
+      <> CBOR.encodeWord 0
+    encode (MsgRequestObjectIds blocking (NumObjectIdsAck ackNo) (NumObjectIdsReq reqNo)) =
+         CBOR.encodeListLen 4
+      <> CBOR.encodeWord 1
+      <> CBOR.encodeBool
+           ( case blocking of
+               SingBlocking    -> True
+               SingNonBlocking -> False
+           )
+      <> CBOR.encodeWord16 ackNo
+      <> CBOR.encodeWord16 reqNo
+    encode (MsgReplyObjectIds objIds) =
+         CBOR.encodeListLen 2
+      <> CBOR.encodeWord 2
+      <> CBOR.encodeListLenIndef
+      <> foldr (\objId r -> encodeObjectId objId <> r) CBOR.encodeBreak objIds'
+           where
+             objIds' :: [objectId]
+             objIds' = case objIds of
+               BlockingReply xs    -> NonEmpty.toList xs
+               NonBlockingReply xs -> xs
```

Comment:
We can use the `Foldable` instance of `BlockingReplyList` and `Monoid` instance of `Encode`:
```suggestion
      <> CBOR.encodeListLenIndef
      <> foldMap encodeObjectId objIds
      <> CBOR.encodeBreak
```